kratos/processes/find_global_nodal_neighbours_for_entities_process.cpp

```diff
@@ -0,0 +1,326 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//                   Suneth Warnakulasuriya
+//
+
+// System includes
+#include <iostream>
+#include <string>
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/element.h"
+#include "includes/global_pointer_variables.h"
+#include "includes/model_part.h"
+#include "includes/node.h"
+#include "processes/process.h"
+#include "utilities/communication_coloring_utilities.h"
+#include "utilities/global_pointer_utilities.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/pointer_communicator.h"
+#include "utilities/variable_utils.h"
+
+// Include base h
+#include "find_global_nodal_neighbours_for_entities_process.h"
+
+namespace Kratos
+{
+template <>
+ModelPart::ElementsContainerType& FindNodalNeighboursForEntitiesProcess<ModelPart::ElementsContainerType>::GetContainer()
+{
+    return this->mrModelPart.Elements();
+}
+
+template <>
+ModelPart::ConditionsContainerType& FindNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType>::GetContainer()
+{
+    return this->mrModelPart.Conditions();
+}
+
+template <>
+void FindNodalNeighboursForEntitiesProcess<ModelPart::ElementsContainerType>::AddHangingNodeIds(
+    std::unordered_map<int, std::unordered_map<int, std::vector<int>>>& rNeighbourIds) const
+{
+    // do nothing in here since mettis partitioner is based on elements, there
+    // cannot be any hanging nodes. If metis partitioner is based on conditions, then
+    // this process need not to be used with elements, so this method won't be called.
+}
+
+template <>
+void FindNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType>::AddHangingNodeIds(
+    std::unordered_map<int, std::unordered_map<int, std::vector<int>>>& rNeighbourIds) const
+{
+    // if the metis partitioner is based on conditions, this will still work,
+    // but with this additional cost of checking.
+    // this->mrModelPart.GetCommunicator().SynchronizeOrNodalFlags(VISITED);
+
+    block_for_each(mrModelPart.Nodes(), [&](const NodeType& rNode) {
+        const int i_owner_rank = rNode.FastGetSolutionStepValue(PARTITION_INDEX);
+        const int node_id = rNode.Id();
+        if (rNeighbourIds[i_owner_rank].find(node_id) ==
+                                     rNeighbourIds[i_owner_rank].end()) {
+#pragma omp critical
```

Comment:
ouch, you are completely killing parallelism by this critical.

at this point i would use a non-parallel for and leave ut the each.

Reply:
actually this is even wrong in parallel: you could be adding something to rNeighbourIds[i_owner_rank] within the critical and at the same time search in the same map.
This is NOT allowed. you would need to have one lock per owner_rank and lock it both at searching time and at adding time. Not sure if doing this in parallel would pay off though...

Reply:
aside of the parallelism problem, i would remove the if and touch it

in short something like

~~~cpp
for(const NodeType& rNode : mrModelPart.Nodes()) {
        const int i_owner_rank = rNode.FastGetSolutionStepValue(PARTITION_INDEX);
        const auto node_id = rNode.Id();
        rNeighbourIds[i_owner_rank][node_id]; //touch directly, avoiding the if
    }
~~~

NOTE take care with the ints with the node id. Better use auto (or Node<3>::IndexType)

Reply:
I was thinking since I am only touhcing `rNeighbourIds[i_owner_rank][node_id]` and if one process is adding `node_id` in critical section and another process is checkig for the same `node_id`, then the second process will also try to add `node_id` which will do nothing for `std::unorderd_maps`. I understand your concern, I was thinking about the small gain. But as you said it feels wrong, I'll change it  as you suggest, it is cleaner :).... thanks @RiccardoRossi  :)